python/ray/data/datasource/parquet_datasource.py

```diff
@@ -254,6 +280,53 @@ def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
 
         return read_tasks
 
+    def _estimate_files_encoding_ratio(self) -> float:
+        """Return an estimate of the Parquet files encoding ratio.
+
+        To avoid OOMs, it is safer to return an over-estimate than an underestimate.
+        """
+        if not DatasetContext.get_current().decoding_size_estimation:
+            return PARQUET_ENCODING_RATIO_ESTIMATE_DEFAULT
+
+        # Sample a few rows from Parquet files to estimate the encoding ratio.
+        # Launch tasks to sample multiple files remotely in parallel.
+        # Evenly distributed to sample N rows in i-th row group in i-th file.
+        # TODO(ekl/cheng) take into account column pruning.
+        start_time = time.perf_counter()
+        num_files = len(self._pq_ds.pieces)
+        num_samples = int(num_files * PARQUET_ENCODING_RATIO_ESTIMATE_SAMPLING_RATIO)
+        min_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MIN_NUM_SAMPLES, num_files
+        )
+        max_num_samples = min(
+            PARQUET_ENCODING_RATIO_ESTIMATE_MAX_NUM_SAMPLES, num_files
+        )
+        num_samples = max(min(num_samples, max_num_samples), min_num_samples)
+
+        # Evenly distributed to choose which file to sample, to avoid biased prediction
+        # if data is skewed.
+        file_samples = [
+            self._pq_ds.pieces[idx]
+            for idx in np.linspace(0, num_files - 1, num_samples).astype(int).tolist()
+        ]
+
+        sample_piece = cached_remote_fn(_sample_piece)
```

Comment:
We expect these file sampling tasks to be I/O-bound and very lightweight, should we make the CPU request relatively lightweight as well to make sure that we're not getting blocked on Ray-level queueing? cc @ericl

Reply:
I think it's OK. Let's see if this becomes a problem --- if it is, we should probably add that and also some progress bars.

Reply:
yes agree with @ericl to add this and progress bars later, if the estimation generally takes a much longer time than we expect. But thanks for @clarkzinzow bringing it up. Beside IO cost, the row group needs to be uncompressed back, so it also costs some CPU resource. In practise, I never compare the ratio between IO vs CPU when reading just a few rows into memory though.